src/python/pants/pantsd/pailgun_server.py

```diff
@@ -178,19 +195,74 @@ def handle_request(self):
     with self.lifecycle_lock():
       self._handle_request_noblock()
 
+  def _should_poll_forever(self, timeout):
+    return timeout < 0
+
+  def _should_keep_polling(self, timeout, time_polled):
+    return self._should_poll_forever(timeout) or time_polled < timeout
+
+  @contextmanager
+  def ensure_request_is_exclusive(self, environment, request):
+    """
+    Ensure that this is the only pants running.
+
+    We currently don't allow parallel pants runs, so this function blocks a request thread until
+    there are no more requests being handled.
+    """
+    # TODO add `did_poll` to pantsd metrics
+
+    timeout = float(environment['PANTSD_REQUEST_TIMEOUT_LIMIT'])
+
+    @contextmanager
+    def yield_and_release(time_waited):
+      try:
+        self.logger.debug("request lock acquired {}.".format("on the first try" if time_waited == 0 else "in {} seconds".format(time_waited)))
+        self.free_to_handle_request_lock.clear()
+        yield
+      finally:
+        self.free_to_handle_request_lock.set()
+        self.logger.debug("released request lock.")
+
+    time_polled = 0.0
+    polling_interval = 1.0 # check the lock every second
```

Comment:
I don't know python's `threading.Event`, but 1s seems like a large polling interval; is there a reason not to drop this by an order of magnitude or two?

Reply:
So the `polling_interval` should probably be called `user_notification_interval`, since we don't actually block for the whole second, we block until _either_ the lock is released or a second has passed, and the notify the user of the timeout has been reached.

So this variable is purely user-facing (to know when to display the "Waiting for 1.0s for a thread to start..." sort of messages), as a request will immediately start when the lock is released.

Will change the name of the variable.